conformance: add test cases for Gateway->Secret ReferenceGrant

nathancoleman opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:

1. If this is your first time contributing to Gateway API, please read our
   developer guide (https://gateway-api.sigs.k8s.io/devguide/)
   and our community page (https://gateway-api.sigs.k8s.io/community/).
2. If this is your first time contributing to a Kubernetes project, please read
   our contributor guidelines:
   https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution
3. Please label this pull request according to what type of issue you are
   addressing, especially if this is a release targeted pull request. For
   reference on required PR/issue labels, read here:
   https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
4. If you want *faster* PR reviews, read how:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design
/kind gep

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature

**What this PR does / why we need it**:
Builds on #1081 to add test coverage for ReferenceGrant as it applied to cross-namespace `Gateway` -> `Secret` references.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #1105

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->
```release-note
Add conformance tests for ReferenceGrant as it applies to cross-namespace Gateway->Secret references
```

nathancoleman commented on conformance/base/manifests.yaml:
```diff
@@ -350,3 +350,20 @@ spec:
+  namespace: secrets
+type: kubernetes.io/tls
+data:
+  tls.crt: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUU2RENDQXRBQ0NRQ2xDM0tMdkJiSlBqQU5CZ2txaGtpRzl3MEJBUXNGQURBMk1Rc3dDUVlEVlFRR0V3SlYKVXpFVE1CRUdBMVVFQ2d3S1MzVmlaWEp1WlhSbGN6RVNNQkFHQTFVRUF3d0piRzlqWVd4b2IzTjBNQjRYRFRJeQpNRFF4T1RFMk5ERXhOMW9YRFRJek1EUXhPVEUyTkRFeE4xb3dOakVMTUFrR0ExVUVCaE1DVlZNeEV6QVJCZ05WCkJBb01Da3QxWW1WeWJtVjBaWE14RWpBUUJnTlZCQU1NQ1d4dlkyRnNhRzl6ZERDQ0FpSXdEUVlKS29aSWh2Y04KQVFF ...
```
This is a self-signed throwaway cert; however, it might trigger scanners somewhere. Any conventions for this that I don't know about?

robscott replied:
I think we may want to create this in the Go test definition instead of a yaml manifest. Maybe we could have a function similar to what was used in ingress-controller-conformance? (https://github.com/kubernetes-sigs/ingress-controller-conformance/blob/cec7cf25b0291933426e256065ef4afc981ea1fa/test/kubernetes/kubernetes.go#L207-L244)

nathancoleman replied:
This turned out to be a good approach, thanks @robscott ! I did need to add a hook to install in-memory objects during `TestSuite.Setup` and have them clean up alongside the yaml manifests that we use for base resources today

nathancoleman commented on conformance/utils/kubernetes/helpers.go:
```diff
@@ -320,6 +320,7 @@ func conditionsMatch(t *testing.T, expected, actual []metav1.Condition) bool {
 func findConditionInList(t *testing.T, conditions []metav1.Condition, condName, condValue string) bool {
 	for _, cond := range conditions {
 		if cond.Type == condName {
+			// TODO(nathancoleman): Compare Reason
```
Still need to address this before marking as ready for review. I suspect we have some tests today that _think_ the `Reason` is being asserted when it actually isn't. Not sure if this will cause some other unrelated tests to start failing.

nathancoleman replied:
This change winds up being non-trivial as all of our test setup assertions break if you start asserting the `Condition.Reason`.

If no one feels strongly, I'd prefer addressing in a followup PR, understanding that it's a pre-existing hole in our test coverage.

Thoughts @robscott ?

robscott replied:
This func only accepts a condition name and value as string right? So it would currently be impossible for any test to expect that a reason was also being checked. I think it's fine to leave checks on reason for a later point.

nathancoleman replied:
Right. I went down the path of accepting a reason arg as well, but plumbing that all the way down without just creating a new set of `...WithReason` functions alongside what we have today is more scope expansion than I'd like in this PR.

nathancoleman commented on conformance/base/manifests.yaml:
```diff
@@ -19,7 +19,7 @@ metadata:
   namespace: gateway-conformance-infra
 spec:
-  gatewayClassName: {GATEWAY_CLASS_NAME}
+  gatewayClassName: "{GATEWAY_CLASS_NAME}"
```
We're still in a position where we need to patch an annotation onto the `Deployments` using `kubectl --kustomize`; however, this was previously always getting mangled as an object instead of a string. It's just a placeholder that gets replaced by the test runner [here](https://github.com/kubernetes-sigs/gateway-api/blob/5a69157e5ca3e7831834af4ca27e3f6ac26db612/conformance/utils/kubernetes/apply.go#L57).

nathancoleman reviewed (commented):
Personal review

skriss reviewed (commented):
Thanks @nathancoleman, had a couple comments from running this against Contour's implementation. Modulo the comments plus one other fix I need to make to Contour, was able to get these to pass.

robscott commented on conformance/tests/gateway-secret-reference-policy.go:
```diff
@@ -0,0 +1,63 @@
+						Type:   string(v1alpha2.ListenerConditionReady),
+						Status: metav1.ConditionTrue,
+					},
+					{
+						Type:   string(v1alpha2.ListenerConditionResolvedRefs),
+						Status: metav1.ConditionTrue,
+					},
```
We haven't been clear about that, my initial thought was to only _require_ conditions like this when they represent an error state. With that said, that would make conformance tests more complex, and it doesn't hurt anything to have a ResolvedRefs condition present... Interested in what others think.

nathancoleman replied:
If they're not currently required in a healthy state, I'm tempted to leave that contract as-is unless it becomes problematic one day. Can adjusts these tests accordingly

nathancoleman commented on conformance/utils/suite/suite.go:
```diff
@@ -124,6 +124,10 @@ func (suite *ConformanceTestSuite) Setup(t *testing.T) {
 	t.Logf("Test Setup: Applying base manifests")
 	suite.Applier.MustApplyWithCleanup(t, suite.Client, suite.BaseManifests, suite.GatewayClassName, suite.Cleanup)
 
+	t.Logf("Test Setup: Applying programmatic resources")
+	secret := kubernetes.MustCreateSelfSignedCertSecret(t, "gateway-conformance-web-backend", "certificate", []string{"*"})
+	suite.Applier.MustApplyObjectsWithCleanup(t, suite.Client, []client.Object{secret}, suite.Cleanup)
```
This is the first resource we're creating for testing that doesn't reside in a yaml file somewhere in the repo, so I needed to add this into the existing setup hook

skriss replied:
It might be nice to be able to attach these kinds of resource creates to the tests for which they're needed rather than having to put them all in the suite-level setup; could add `Setup`/`Teardown`-style functions to the `ConformanceTest` struct. I don't think this needs to block now, but could be a refactor to keep in mind if/when we add more of these.

nathancoleman commented on conformance/utils/suite/suite.go:
```diff
@@ -124,6 +124,10 @@ func (suite *ConformanceTestSuite) Setup(t *testing.T) {
 	suite.Applier.MustApplyWithCleanup(t, suite.Client, suite.BaseManifests, suite.GatewayClassName, suite.Cleanup)
 
+	t.Logf("Test Setup: Applying programmatic resources")
+	secret := kubernetes.MustCreateSelfSignedCertSecret(t, "gateway-conformance-web-backend", "certificate", []string{"*"})
```
The `*` hostname might stick out, but the changes made in this PR only cover ReferencePolicy. We may want tests in the future that cover the request-response when using a certificate; however, that is outside the scope of this PR.

nathancoleman commented:
@skriss @robscott please review latest commits when ya get a chance

nathancoleman commented:
@skriss I've added many different permutations for inapplicable ReferencePolicies to the test that already existed, asserting that none of them apply.

I've also added a new case w/ a ReferencePolicy granting access to all Secrets in the target namespace vs. a specific one. All are passing for Consul. Please confirm for Contour and refresh your review when you get a chance.

robscott commented:
/ok-to-test

skriss commented:
> @skriss I've added many different permutations for inapplicable ReferencePolicies to the test that already existed, asserting that none of them apply.
> 
> I've also added a new case w/ a ReferencePolicy granting access to all Secrets in the target namespace vs. a specific one. All are passing for Consul. Please confirm for Contour and refresh your review when you get a chance.

Nice! All are passing for Contour. I'll take a look through the code shortly.

youngnick commented:
/lgtm

Nice work @nathancoleman!

nathancoleman commented:
@robscott who's responsible for providing the approval to merge?

robscott commented:
🤦 Somehow I'd thought Nick had approved and I was trying to do the other half.

/approve
